### PR TITLE
*: adjust TLS behaviour for dumpling and lightning

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/37480

Problem Summary:

### What is changed and how it works?

align the behaviour of dumpling/lightning with MySQL client

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

#### dumpling

dumpling with only CA:

```
$ mysql -P3306 -h127.0.0.1 -uroot -p123456 --ssl-ca=/tmp/server-ca.pem -e "SELECT version()" 
+------------+
| version()  |
+------------+
| 5.7.38-log |
+------------+

$ bin/dumpling -h 127.0.0.1 --port 3306 --password 123456 --ca /tmp/server-ca.pem -B tls
Release version: v6.3.0-alpha-118-g7a956244b-dev
Git commit hash: 7a956244bdd5e3f5d7f5ed15f8e162cea6ea3648
Git branch:      fix-check-ca
Build timestamp: 2022-09-05 06:30:05Z
Go version:      go version go1.19 linux/amd64

[2022/09/05 14:31:00.699 +08:00] [INFO] [versions.go:54] ["Welcome to dumpling"] ["Release Version"=v6.3.0-alpha-118-g7a956244b-dev] ["Git Commit Hash"=7a956244bdd5e3f5d7f5ed15f8e162cea6ea3648] ["Git Branch"=fix-check-ca] ["Build timestamp"="2022-09-05 06:30:05"] ["Go Version"="go version go1.19 linux/amd64"]
...
[2022/09/05 14:31:00.727 +08:00] [INFO] [main.go:81] ["dump data successfully, dumpling will exit now"]
```

dumpling with only CA but wrong CA

```
$ mysql -P3307 -h127.0.0.1 -uroot -p123456 --ssl-ca=/tmp/server-ca.pem -e "SELECT version()"  
ERROR 2026 (HY000): SSL connection error: self signed certificate in certificate chain

$ bin/dumpling -h 127.0.0.1 --port 3307 --password 123456 --ca /tmp/server-ca.pem -B tls -o /tmp/dumpling_test
Release version: v6.3.0-alpha-118-g7a956244b-dev
Git commit hash: 7a956244bdd5e3f5d7f5ed15f8e162cea6ea3648
Git branch:      fix-check-ca
Build timestamp: 2022-09-05 06:30:05Z
Go version:      go version go1.19 linux/amd64

[2022/09/05 14:34:37.744 +08:00] [INFO] [versions.go:54] ["Welcome to dumpling"] ["Release Version"=v6.3.0-alpha-118-g7a956244b-dev] ["Git Commit Hash"=7a956244bdd5e3f5d7f5ed15f8e162cea6ea3648] ["Git Branch"=fix-check-ca] ["Build timestamp"="2022-09-05 06:30:05"] ["Go Version"="go version go1.19 linux/amd64"]
[2022/09/05 14:34:37.746 +08:00] [WARN] [version.go:299] ["select tidb_version() failed, will fallback to 'select version();'"] [error="can't verify certificate, maybe different CA is used: x509: certificate signed by unknown authority"]

create dumper failed: sql: SELECT version();: can't verify certificate, maybe different CA is used: x509: certificate signed by unknown authority
```

dumpling with only client credentials

```
$ mysql -P3306 -h127.0.0.1 -uroot -p123456 --ssl-cert /tmp/tiflow_engine_test/dm_tls/mysql1/client-cert.pem --ssl-key /tmp/tiflow_engine_test/dm_tls/mysql1/client-key.pem -e "SELECT version()"
+------------+
| version()  |
+------------+
| 5.7.38-log |
+------------+

$ bin/dumpling -h 127.0.0.1 --port 3306 --password 123456 --cert /tmp/tiflow_engine_test/dm_tls/mysql1/client-cert.pem --key /tmp/tiflow_engine_test/dm_tls/mysql1/client-key.pem  -B tls -o /tmp/dumpling_test
Release version: v6.3.0-alpha-118-g7a956244b-dev
Git commit hash: 7a956244bdd5e3f5d7f5ed15f8e162cea6ea3648
Git branch:      fix-check-ca
Build timestamp: 2022-09-05 06:30:05Z
Go version:      go version go1.19 linux/amd64

[2022/09/05 14:39:46.940 +08:00] [INFO] [versions.go:54] ["Welcome to dumpling"] ["Release Version"=v6.3.0-alpha-118-g7a956244b-dev] ["Git Commit Hash"=7a956244bdd5e3f5d7f5ed15f8e162cea6ea3648] ["Git Branch"=fix-check-ca] ["Build timestamp"="2022-09-05 06:30:05"] ["Go Version"="go version go1.19 linux/amd64"]
...
[2022/09/05 14:39:46.974 +08:00] [INFO] [main.go:81] ["dump data successfully, dumpling will exit now"]
```

dumpling with only client credentials but wrong

```
$ mysql -P3307 -h127.0.0.1 -uroot -p123456 --ssl-cert /tmp/tiflow_engine_test/dm_tls/mysql1/client-cert.pem --ssl-key /tmp/tiflow_engine_test/dm_tls/mysql1/client-key.pem -e "SELECT version()"
ERROR 2013 (HY000): Lost connection to server during query

$ bin/dumpling -h 127.0.0.1 --port 3307 --password 123456 --cert /tmp/tiflow_engine_test/dm_tls/mysql1/client-cert.pem --key /tmp/tiflow_engine_test/dm_tls/mysql1/client-key.pem  -B tls -o /tmp/dumpling_test
Release version: v6.3.0-alpha-118-g7a956244b-dev
Git commit hash: 7a956244bdd5e3f5d7f5ed15f8e162cea6ea3648
Git branch:      fix-check-ca
Build timestamp: 2022-09-05 06:30:05Z
Go version:      go version go1.19 linux/amd64

[2022/09/05 14:42:10.310 +08:00] [INFO] [versions.go:54] ["Welcome to dumpling"] ["Release Version"=v6.3.0-alpha-118-g7a956244b-dev] ["Git Commit Hash"=7a956244bdd5e3f5d7f5ed15f8e162cea6ea3648] ["Git Branch"=fix-check-ca] ["Build timestamp"="2022-09-05 06:30:05"] ["Go Version"="go version go1.19 linux/amd64"]
[mysql] 2022/09/05 14:42:10 packets.go:37: remote error: tls: unknown certificate authority
[2022/09/05 14:42:10.313 +08:00] [WARN] [version.go:299] ["select tidb_version() failed, will fallback to 'select version();'"] [error="invalid connection"]
[mysql] 2022/09/05 14:42:10 packets.go:37: remote error: tls: unknown certificate authority

create dumper failed: sql: SELECT version();: invalid connection
```

#### lightning

lightning with only CA

```
$ mysql -P4000 -h127.0.0.1 -uroot --ssl-ca=/tmp/certs/downstream/ca.pem -e "SELECT version()" 
+--------------------+
| version()          |
+--------------------+
| 5.7.25-TiDB-v5.4.0 |
+--------------------+

$ bin/tidb-lightning -d /tmp/dumpling_test -backend tidb -ca /tmp/certs/downstream/ca.pem -tidb-port 4000
Verbose debug logs will be written to /tmp/lightning.log.2022-09-05T15.42.17+0800
...
tidb lightning exit successfully
```

lightning with only CA but wrong CA

```
$ mysql -P4000 -h127.0.0.1 -uroot -p123456 --ssl-ca=/tmp/server-ca.pem -e "SELECT version()"
ERROR 2026 (HY000): SSL connection error: unable to get local issuer certificate

$ bin/tidb-lightning -d /tmp/dumpling_test -backend tidb -ca /tmp/server-ca.pem -tidb-password 123456 -tidb-port 4000
Verbose debug logs will be written to /tmp/lightning.log.2022-09-05T15.38.51+0800

tidb lightning encountered error: [Lightning:DB:ErrDBConnect]failed to connect database: can't verify certificate, maybe different CA is used: x509: certificate signed by unknown authority
```

lightning with only client credentials

```
$ mysql -P4000 -h127.0.0.1 -udm_user --ssl-cert=/tmp/certs/downstream/client.pem --ssl-key=/tmp/certs/downstream/client.key -e "SELECT version()"
+--------------------+
| version()          |
+--------------------+
| 5.7.25-TiDB-v5.4.0 |
+--------------------+

$ bin/tidb-lightning -d /tmp/dumpling_test -backend tidb -tidb-user dm_user -tidb-port 4000 -cert /tmp/certs/downstream/client.pem -key /tmp/certs/downstream/client.key
Verbose debug logs will be written to /tmp/lightning.log.2022-09-05T16.05.23+0800
...
tidb lightning exit successfully
```

lightning with only client credentials but wrong

```
$ mysql -P4000 -h127.0.0.1 -udm_user --ssl-cert=/tmp/tiflow_engine_test/dm_tls/mysql1/client-cert.pem --ssl-key=/tmp/tiflow_engine_test/dm_tls/mysql1/client-key.pem  -e "SELECT version()"
ERROR 2013 (HY000): Lost connection to server at 'reading authorization packet', system error: 0

$ bin/tidb-lightning -d /tmp/dumpling_test -backend tidb -tidb-user dm_user -tidb-port 4000 -cert /tmp/tiflow_engine_test/dm_tls/mysql1/client-cert.pem -key /tmp/tiflow_engine_test/dm_tls/mysql1/client-key.pem
Verbose debug logs will be written to /tmp/lightning.log.2022-09-05T16.08.37+0800
...
tidb lightning encountered error: [Lightning:DB:ErrDBConnect]failed to connect database: Error 1045: Access denied for user 'dm_user'@'192.168.208.1' (using password: NO)
```

- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by lance6716** [comment]
/hold

suspect our verify cn logic is not covered

**Comment by lance6716** [comment]
ptal @dsdashun @overvenus

**Comment by lance6716** [comment]
ptal @Ehco1996 @gozssky

**Comment by dveeden** [comment]
Did you try with `mysql --tls-mode=VERIFY_IDENTITY ...` and the equivalent options or dumpling/lightning?
What about `--tls-version` ?
Is there an equivalent of `mysql --tls-mode=REQUIRED ...`?

**Comment by lance6716** [comment]
> Did you try with `mysql --tls-mode=VERIFY_IDENTITY ...` and the equivalent options or dumpling/lightning? What about `--tls-version` ? Is there an equivalent of `mysql --tls-mode=REQUIRED ...`?

I uses an old MySQL(MariaDB) client it doesn't not support `--tls-mode` :joy:

```
$ mysql --version
mysql  Ver 15.1 Distrib 10.8.3-MariaDB, for Linux (x86_64) using readline 5.1
```

And I learned current behaviour should be same as VERIFY_CA

> If [--ssl-mode](https://dev.mysql.com/doc/refman/5.7/en/connection-options.html#option_general_ssl-mode) is not explicitly set otherwise, use of [--ssl-ca](https://dev.mysql.com/doc/refman/5.7/en/connection-options.html#option_general_ssl-ca) or [--ssl-capath](https://dev.mysql.com/doc/refman/5.7/en/connection-options.html#option_general_ssl-capath) implies [--ssl-mode=VERIFY_CA](https://dev.mysql.com/doc/refman/5.7/en/connection-options.html#option_general_ssl-mode).

And we support TLSv1 as minimum tls to keep compatibility with MySQL 5.7

File: util/security.go [hunk 1]
```diff
@@ -107,34 +107,94 @@ func ToTLSConfigWithVerify(caPath, certPath, keyPath string, verifyCN []string)
...
 	/* #nosec G402 */
 	tlsCfg := &tls.Config{
-		MinVersion:   tls.VersionTLS10,
-		Certificates: certificates,
-		RootCAs:      certPool,
-		ClientCAs:    certPool,
-		NextProtos:   []string{"h2", "http/1.2"}, // specify `h2` to let Go use HTTP/2.
+		MinVersion:         tls.VersionTLS10,
+		Certificates:       certificates,
+		InsecureSkipVerify: true,
```
[the remark this answers is not in the document]
Comment (lance6716) [reply util/security.go:136]:
```
	// InsecureSkipVerify controls whether a client verifies the server's
	// certificate chain and host name. If InsecureSkipVerify is true, crypto/tls
	// accepts any certificate presented by the server and any host name in that
	// certificate. In this mode, TLS is susceptible to machine-in-the-middle
	// attacks unless custom verification is used. This should be used only for
	// testing or in combination with VerifyConnection or VerifyPeerCertificate.
	InsecureSkipVerify bool
```

**Review by lichunzhu — APPROVED** [review]
LGTM

File: util/security.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dveeden) [reply util/security.go:136]: There are two things to verify:

- Is the certificate signed by one of the trusted CA's? (like `--ssl-mode=VERIFY_CA`)
- Does the commonName or subjectAlternativeName in the certiicate match the host. (like `--ssl-mode=VERIFY_IDENTITY`)

Did we check both with `VerifyPeerCertificate`?

**Review by dveeden — APPROVED** [review]

**Comment by lance6716** [comment]
/hold

need to address comment and wait review for lightning ceritificate rotation check

File: util/security.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dveeden) [reply util/security.go:136]: Only validating the certificate is often used when loadbalancers etc. are used and when the certificate that the server provides doesn't contain the hostname of the loadbalancer in the subjectAlternativeName. This is only required when the loadbalancer doesn't do TLS termination.

Having the client validate the cert and hostname is the only really secure option.

I assume that things like `Not After`/`Not Before` and other important TLS/X.509 things are checked by Go in either case.

**Review by gozssky — APPROVED** [review]

**Pull request merged**